### PR TITLE
[Infra] Specialise Initialisation functions for team experiments

### PR DESCRIPTION
**Pull request by sherwin-dc**
- Teams can now create their own agent map within their team files for
    experimentation.
- Refactor initialise code out of main
- Added an example probabilistic agent in team0 files

### CODE REVIEW

**Comment by 0x6770** [comment]
I thought we have already enabled customisable agent map via #25?

**Comment by neeldug** [comment]
> I thought we have already enabled customisable agent map via #25?

I thought so too…

**Comment by sherwin-dc** [comment]
This has nothing to do with mixing different agent types. This PR allows for the agent map to be kept in another file eg `infra/teams/team0/init.go`, otherwise each team would be keeping a local copy of the whole repo with their own agent map in `main.go`

**Comment by 0x6770** [comment]
> This has nothing to do with mixing different agent types. This PR allows for the agent map to be kept in another file eg `infra/teams/team0/init.go`, otherwise each team would be keeping a local copy of the whole repo with their own agent map in `main.go`

oh I see, if you want to add a way to allow teams/members to save their agent map somewhere, I would suggest to add a flag to config `env` path. By this way, we don't need to modify the existing code, but only add one line in `main.go`

**Comment by 0x6770** [comment]
Something like this
```
diff --git a/pkg/infra/main.go b/pkg/infra/main.go
index 27badf4..e6bc282 100644
--- a/pkg/infra/main.go
+++ b/pkg/infra/main.go
@@ -35,11 +35,12 @@ func main() {
        // define flags
        useJSONFormatter := flag.Bool("j", false, "Whether to output logs in JSON")
        debug := flag.Bool("d", false, "Whether to run in debug mode. If false, only logs with level info or above will be shown")
+       envFile := flag.String("c", ".env", "Path to environment file, default to .env")
        flag.Parse()

        logging.InitLogger(*useJSONFormatter, *debug)

-       agentMap, globalState, gameConfig := initialise()
+       agentMap, globalState, gameConfig := initialise(*envFile)
        gameLoop(globalState, agentMap, gameConfig)
 }

@@ -110,11 +111,11 @@ func gameLoop(globalState state.State, agentMap map[commons.ID]agent.Agent, game
        logging.Log(logging.Info, nil, fmt.Sprintf("Congratulations, The Peasants have escaped the pit with %d remaining.", len(agentMap)))
 }

-func initialise() (map[commons.ID]agent.Agent, state.State, config.GameConfig) {
+func initialise(envFile string) (map[commons.ID]agent.Agent, state.State, config.GameConfig) {
        agentMap := make(map[commons.ID]agent.Agent)
        agentStateMap := make(map[commons.ID]state.AgentState)

-       err := godotenv.Load()
+       err := godotenv.Load(envFile)
        if err != nil {
                logging.Log(logging.Error, nil, "No .env file located, using defaults")
        }
```

**Comment by neeldug** [comment]
> > This has nothing to do with mixing different agent types. This PR allows for the agent map to be kept in another file eg `infra/teams/team0/init.go`, otherwise each team would be keeping a local copy of the whole repo with their own agent map in `main.go`
> 
> oh I see, if you want to add a way to allow teams/members to save their agent map somewhere, I would suggest to add a flag to config `env` path. By this way, we don't need to modify the existing code, but only add one line in `main.go`

Yep I agree with this — since in reality the env files enable full customisability.

**Comment by sherwin-dc** [comment]
Yes that does seem simpler, but do you know how to load string values into a map to fit its types? If we have a map of `string` to `string` it would be easy to load values from a json file and I can find examples online, but in this case how would we load text into a map of `string` to `agent.Strategy` func?

Lets say the json/env file contains something like
```
	"RANDOM":          agent.NewRandomAgent(),
	"AggressiveAgent": NewProbabilisticAgent(0.1, 0.8, 0.1),
	"CowardlyAgent":   NewProbabilisticAgent(0.9, 0.05, 0.05),
```

**Comment by 0x6770** [comment]
```
# in .env_team0
MODE=default
LEVELS=60
STARTING_HP=1000
STARTING_ATTACK=20
STARTING_SHIELD=20
BASE_STAMINA=2000
THRESHOLD_PCT=0.6
AGENT_RANDOM_QUANTITY=100
AGENT_AggressiveAgent_QUANTITY=20
AGENT_CowardlyAgent_QUANTITY=20
```

```go
// in team0_agent.go
func NewAggressiveAgent() = NewProbabilisticAgent(0.1, 0.8, 0.1)
func NewCowardlyAgent() = NewProbabilisticAgent(0.9, 0.05, 0.05)
```

```go
// in main.go
var InitAgentMap = map[commons.ID]agent.Strategy{
	"RANDOM": agent.NewRandomAgent(),
        "AggressiveAgent": team0.NewAggressiveAgent(),
        "CowardlyAgent": team0.NewCowardlyAgent(),
}
```

**Comment by sherwin-dc** [comment]
> ```
> # in .env_team0
> MODE=default
> LEVELS=60
> STARTING_HP=1000
> STARTING_ATTACK=20
> STARTING_SHIELD=20
> BASE_STAMINA=2000
> THRESHOLD_PCT=0.6
> AGENT_RANDOM_QUANTITY=100
> AGENT_AggressiveAgent_QUANTITY=20
> AGENT_CowardlyAgent_QUANTITY=20
> ```
> 
> ```go
> // in team0_agent.go
> func NewAggressiveAgent() = NewProbabilisticAgent(0.1, 0.8, 0.1)
> func NewCowardlyAgent() = NewProbabilisticAgent(0.9, 0.05, 0.05)
> ```
> 
> ```go
> // in main.go
> var InitAgentMap = map[commons.ID]agent.Strategy{
> 	"RANDOM": agent.NewRandomAgent(),
>         "AggressiveAgent": team0.NewAggressiveAgent(),
>         "CowardlyAgent": team0.NewCowardlyAgent(),
> }
> ```

But this means each team would need to still write into the same agent map in main.go

**Comment by sherwin-dc** [comment]
One of the purposes of my PR is to have teams define their own `var InitAgentMap = map[commons.ID]agent.Strategy{ ... }` within their files for their experiments

**Comment by 0x6770** [comment]
But this means every time other teams add new agent, you will need to modify that map if you want to include that new agent in your run.

**Comment by Razvan00Rusu** [comment]
I havent had the time to look over this pr properly yet - but ill get home soon so can do then. 

But just from the conversation here I think it does make sense to move the agent map to agent teams as (I presume) proposed and maybe selected via a flag to avoid conflicts in main.go when teams merge their code in

**Comment by sherwin-dc** [comment]
> But just from the conversation here I think it does make sense to move the agent map to agent teams as (I presume) proposed and maybe selected via a flag to avoid conflicts in main.go when teams merge their code in

Yes this, though I intended it for team experiments. It should be straightforward to extend this to combine the maps defined in multiple files so that teams can write their own maps even for normal gameplay against other teams.

> But this means every time other teams add new agent, you will need to modify that map if you want to include that new agent in your run.

This is intended for private team experiments where runs only have agents from one specific team. In this PR the `var InitAgentMap = map[commons.ID]agent.Strategy{ ... }` in main.go is still preserved and functions the same way (though we could follow @Razvan00Rusu's suggestion to also split it up if its desirable).

File: pkg/infra/game/stage/initialise/initialise.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		StartingShieldStrength: config.EnvToUint("STARTING_SHIELD", 100),
+		ThresholdPercentage:    config.EnvToFloat("THRESHOLD_PCT", 0.6),
+		InitialNumAgents:       uint(0),
+		Stamina:                config.EnvToUint("BASE_STAMINA", 2000),
+	}
+
+	return gameConfig
+}
+
+func InitAgents(defaultStratergyMap map[commons.ID]agent.Strategy, gameConfig config.GameConfig) (numAgents uint, agentMap map[commons.ID]agent.Agent, agentStateMap map[commons.ID]state.AgentState) {
```
Comment (Razvan00Rusu) [inline pkg/infra/game/stage/initialise/initialise.go:52]:
```suggestion
func InitAgents(defaultStrategyMap map[commons.ID]agent.Strategy, gameConfig config.GameConfig) (numAgents uint, agentMap map[commons.ID]agent.Agent, agentStateMap map[commons.ID]state.AgentState) {
```

File: pkg/infra/game/stage/initialise/initialise.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	agentMap = make(map[commons.ID]agent.Agent)
+	agentStateMap = make(map[commons.ID]state.AgentState)
+
+	numAgents = 0
+
+	for agentName, strategy := range defaultStratergyMap {
```
Comment (Razvan00Rusu) [inline pkg/infra/game/stage/initialise/initialise.go:58]:
```suggestion
	for agentName, strategy := range defaultStrategyMap {
```

File: pkg/infra/game/stages/stages.go [hunk 3]
```diff
@@ -15,9 +17,36 @@ import (
 	t0 "infra/teams/team0"
 )
 
-// ? Changed at compile time. eg run with `make TEAM=0` to set this to '0'
+// ? Changed at compile time. eg change in .env to `MODE=0` to set this to '0'
 var Mode string
 
+func ChooseDefaultStratergyMap(defaultStratergyMap map[commons.ID]agent.Strategy) map[commons.ID]agent.Strategy {
```
Comment (Razvan00Rusu) [inline pkg/infra/game/stages/stages.go:23]:
```suggestion
func ChooseDefaultStratergyMap(defaultStrategyMap map[commons.ID]agent.Strategy) map[commons.ID]agent.Strategy {
```

File: pkg/infra/game/stages/stages.go [hunk 4]
```diff
@@ -15,9 +17,36 @@ import (
...
+	switch Mode {
+	case "0":
+		return t0.InitAgentMap
+	default:
+		return defaultStratergyMap
```
Comment (Razvan00Rusu) [inline pkg/infra/game/stages/stages.go:28]:
```suggestion
		return defaultStrategyMap
```

**Review by Razvan00Rusu — CHANGES_REQUESTED** [review]
A small typo in 2 of the files but other than that looks good to me

**Review by Razvan00Rusu — APPROVED** [review]

**Comment by sherwin-dc** [comment]
Should we wait for any more comments or is it ready to merge in?

**Comment by Razvan00Rusu** [comment]
> Should we wait for any more comments or is it ready to merge in?

Merge merge  mergeeee

**Pull request merged**